src/search/executors/hnsw_vector_field_knn_scan_executor.h

```diff
@@ -0,0 +1,76 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+
+#pragma once
+
+#include <string>
+
+#include "db_util.h"
+#include "encoding.h"
+#include "search/hnsw_indexer.h"
+#include "search/plan_executor.h"
+#include "search/search_encoding.h"
+#include "storage/redis_db.h"
+#include "storage/redis_metadata.h"
+#include "storage/storage.h"
+#include "string_util.h"
+
+namespace kqir {
+
+// TODO(Beihao): Add DB context to improve consistency and isolation - see #2332
+struct HnswVectorFieldKnnScanExecutor : ExecutorNode {
+  HnswVectorFieldKnnScan *scan;
+  redis::LatestSnapShot ss;
+  bool initialized = false;
+
+  IndexInfo *index;
+  redis::SearchKey search_key;
+  redis::HnswVectorFieldMetadata field_metadata;
```

Comment:
seems here we can just keep a pointer e.g. `redis::HnswVectorFieldMetadata *field_metadata`.

Reply:
Oh this was because `HnswIndex` constructor does not support `const HnswVectorFieldMetadata*` . 
```
HnswIndex(const SearchKey& search_key, HnswVectorFieldMetadata* vector, engine::Storage* storage);
```

`HnswIndex` needs to modify `HnswVectorFieldMetadata*`  in other functions, so I did copy here. Do you have ideas on how we can improve it?

Reply:
Let me think, if the metadata is modified in a certain vector query, it seems that the global metadata of the vector field (in `server->index_mgr->index_map`) will not change. Will this cause any problems?

Reply:
`HnswVectorFieldMetadata*` is only modified when the node is inserted/deleted in a higher layer than all other nodes, so `num_levels` in `HnswVectorFieldMetadata : IndexFieldMetadata` is modified. So the affected field is `server->index_mgr->index_map[<index_key>]->fields[<field>]->metadata`. 

In `indexer.cc`, we do 
```
auto *metadata = iter->second.metadata.get();
if (auto vector = dynamic_cast<HnswVectorFieldMetadata *>(metadata)) {
    GET_OR_RET(UpdateHnswVectorIndex(key, original, current, search_key, vector));
}
```
So `*metadata` changes correspondingly. 

But in this PR, since we don't modify `metadata`, I guess it will not cause actual problems by copying (maybe with consistency issue but this would be fixed later aligning with #2310 ).

Reply:
I see what you mean.

It's good for now, but we can enhance it later. For example, we could create a const version of HnswIndex that takes a const pointer to HnswVectorFieldMetadata.

Reply:
That sounds good!
I wanted to overload with a const version, but there are some nested calls also asking for `HnswVectorFieldMetadata*` as a parameter, so eventually didn't implement it to avoid making this PR look too messy. I'll take a note in tracking issue #2426 and improve it in the future.